DAOS-14181 control: Remove bdev scan cache

tanabarr opened the pull request:
Instead of storing controller details in cache before starting engines,
reread info when scanning over dRPC. This enables details of
replacement controller to be returned after a hotplug event and
reduces some control-plane complexity.

Required-githooks: true

### Before requesting gatekeeper:

* [x] Two review approvals and any prior change requests have been resolved.
* [x] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [x] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [x] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [x] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [x] You are the appropriate gatekeeper to be landing the patch.
* [x] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [x] Githooks were used. If not, request that user install them and check copyright dates.
* [x] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [x] All builds have passed.  Check non-required builds for any new compiler warnings.
* [x] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [x] If applicable, the PR has addressed any potential version compatibility issues.
* [x] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [x] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Build on Leap 15.4 with Intel-C and TARGET_PREFIX completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13385/1/execution/node/393/log

daosbuild1 commented:
Test stage Build on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13385/1/execution/node/411/log

daosbuild1 commented:
Test stage Build RPM on EL 9 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13385/1/execution/node/383/log

daosbuild1 commented:
Test stage Build RPM on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13385/1/execution/node/379/log

daosbuild1 commented:
Test stage Build RPM on Leap 15.4 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13385/1/execution/node/356/log

daosbuild1 commented:
Test stage Build DEB on Ubuntu 20.04 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13385/1/execution/node/362/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Build on Leap 15.4 with Intel-C and TARGET_PREFIX completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13385/2/execution/node/373/log

daosbuild1 commented:
Test stage Build on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13385/2/execution/node/361/log

daosbuild1 commented:
Test stage Build RPM on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13385/2/execution/node/411/log

daosbuild1 commented:
Test stage Build RPM on Leap 15.4 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13385/2/execution/node/403/log

daosbuild1 commented:
Test stage Build DEB on Ubuntu 20.04 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13385/2/execution/node/367/log

daosbuild1 commented:
Test stage NLT on EL 8 completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-13385/3/testReport/

kjacque commented on src/control/server/storage/provider_test.go:
```diff
@@ -33,304 +31,304 @@ func defBdevCmpOpts() []cmp.Option {
-		})
-	}
-}
+//func Test_scanBdevsTiers(t *testing.T) {
```
Needs removal?

tanabarr replied:
done

kjacque commented on src/control/server/storage/bdev_test.go:
```diff
@@ -167,32 +168,48 @@ func Test_Convert_SmdDevice(t *testing.T) {
-		`"serial":"","pci_addr":"0000:01:00.0","fw_rev":"","vendor_id":"","pci_type":"",` +
-		`"socket_id":0,"health_stats":null,"namespaces":null,"smd_devices":null,` +
-		`"dev_state":"EVICTED","led_state":"ON"},"ctrlr_namespace_id":0}`
+	expOut := `{"role_bits":7,"uuid":"00000001-0001-0001-0001-000000000001","` +
```
Possible to break this up on word boundaries? Seems like it would make codespell happier, and it'd be a bit easier for humans to read.

tanabarr replied:
tbh don't think that's an efficient use of time, would prefer not to unless you insist

kjacque replied:
I just hate seeing red on the GHA. It's minor though, wouldn't block on it.

kjacque commented on src/control/server/instance_storage_test.go:
```diff
@@ -1,5 +1,5 @@
 //
-// (C) Copyright 2020-2022 Intel Corporation.
+// (C) Copyright 2020-2023 Intel Corporation.
```
Revert? Looks like no code changes

tanabarr replied:
done

kjacque commented on src/control/server/instance_storage_rpc_test.go:
```diff
@@ -0,0 +1,192 @@
+				State: new(ctlpb.ResponseState),
+			},
+		},
+		// "scan over drpc; with smd": {},
+		// "scan over drpc; with smd and health": {},
```
Test cases to be added?

kjacque commented on src/control/server/instance_storage_rpc.go:
```diff
@@ -157,20 +162,121 @@ func (ei *EngineInstance) StorageFormatNVMe() (cResults proto.NvmeControllerResu
+		if pbReq.Meta {
+			nsd := new(ctlpb.SmdDevice)
+			*nsd = *sd
+			nsd.Ctrlr, nsd.MetaSize, nsd.RdbSize = nil, pbReq.MetaSize, pbReq.RdbSize
```
I know this is legal Go, but this multiple-assignment line is sort of confusing to read. Could it be broken up into separate assignments?

tanabarr replied:
sure

kjacque commented on src/control/server/ctl_storage_rpc.go:
```diff
@@ -61,73 +70,118 @@ func newResponseState(inErr error, badStatus ctlpb.ResponseStatus, infoMsg strin
+
+	pbCtrlrs := make(proto.NvmeControllers, 0, len(resp.Controllers))
+
+	return &ctlpb.ScanNvmeResp{
+		State:  new(ctlpb.ResponseState),
+		Ctrlrs: pbCtrlrs,
+	}, pbCtrlrs.FromNative(resp.Controllers)
```
Should we be returning the resp even if there's an error?

tanabarr replied:
I don't really see the problem with doing so given we check for error before using a response in all the callers.

tanabarr replied:
done in the next commit

mjmac replied:
It is certainly best practice to check that the returned error was non-nil before using the response, but best practices aren't always followed. Better to return a nil response for errors so that any programming mistake is detected quickly during development (hopefully!).

kjacque commented on src/control/server/ctl_storage_rpc.go:
```diff
@@ -61,73 +70,118 @@ func newResponseState(inErr error, badStatus ctlpb.ResponseStatus, infoMsg strin
+	cfgBdevs := getBdevCfgsFromSrvCfg(cs.srvCfg).Bdevs()
+	hasBdevs := cfgBdevs.Len() != 0
+
+	// Note the potential window where engines are started but not yet ready to respond. In this
+	// state there is a possibility that neither scan mechanism will work because devices have
+	// been claimed by SPDK but details are not yet available over dRPC. Here it is assumed that
+	// devices are claimed as soon the engines are started.
```
What happens in this case? Does the drpc request return -DER_AGAIN?

kjacque replied:
I see potentially another race, where the code here sees no engines started, but then the engines start before the global bdev scan executes. What does that case look like to the user?

tanabarr replied:
1. In the situation where SPDK has claimed the controllers for the engine process but dRPC is not yet ready for comms, an "engine not ready" error should be returned. The window should be <<1s.
2. In the situation where this check results in a global bdev scan but by the time the scan executes the controllers have been claimed by the engine process, I think no results will be returned.

In terms of how to handle the above 2 scenarios I suggest the following:
- If "engine not ready" we issue a similar fault suggesting a retry.
- If no bdevs are found but some are expected (as bdev_list has been populated in config), fault suggesting retry is given.

I would prefer not to make it more complicated than this at the current stage to avoid the risk of pre-empting a solution for something that isn't a problem in practice.
Does the above suggestion sound acceptable for now?

kjacque replied:
Sounds good to me. I just wanted to make sure we weren't hiding an error or returning weird/confusing results.

mjmac replied:
I think that Kris raises a valid concern, particularly in the context of CI testing -- we don't want to introduce a new source of intermittent failures. Why should the retry error be exposed to the user vs. just handling the retry in code?

tanabarr replied:
Changes to implement retries when scanning over provider and no results are returned and also to wait for drpc to be ready before scanning when engine is running have been applied in this PR.

kjacque commented on src/control/server/ctl_storage_rpc_test.go:
```diff
@@ -276,2055 +648,81 @@ func TestServer_CtlSvc_StorageScan_PreEngineStart(t *testing.T) {
-//	}
-//}
-
-func TestServer_checkTmpfsMem(t *testing.T) {
```
This wasn't meant to be removed, was it?

tanabarr replied:
nope, good catch, restored in the next commit, thanks

daosbuild1 commented:
Test stage Functional Hardware Large completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-13385/3/testReport/

tanabarr commented on src/control/server/ctl_storage_rpc_test.go:
```diff
@@ -276,2055 +648,81 @@ func TestServer_CtlSvc_StorageScan_PreEngineStart(t *testing.T) {
-	}
-}
-
-func TestServer_CtlSvc_StorageFormat(t *testing.T) {
```
nope, good catch, restored in the next commit, thanks

daosbuild1 commented:
Test stage NLT on EL 8 completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-13385/4/testReport/

daosbuild1 commented:
Test stage Functional Hardware Large completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-13385/4/testReport/

knard-intel reviewed (approved):
LGTM

daosbuild1 commented:
Test stage Functional Hardware Medium Verbs Provider completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13385/4/execution/node/1521/log

daosbuild1 commented:
Test stage Functional Hardware Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13385/4/execution/node/1429/log

daosbuild1 commented:
Test stage Functional Hardware Medium MD on SSD completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13385/4/execution/node/1475/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage NLT on EL 8 completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-13385/7/testReport/

daosbuild1 commented:
Test stage Functional Hardware Large completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-13385/7/testReport/

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage NLT on EL 8 completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-13385/8/testReport/

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage NLT on EL 8 completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-13385/9/testReport/

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

kjacque commented on src/mgmt/srv_query.c:
```diff
@@ -761,7 +761,23 @@ ds_mgmt_dev_manage_led(Ctl__LedManageReq *req, Ctl__DevManageResp *resp)
+	D_ALLOC(resp->device->ctrlr->pci_addr, ADDR_STR_MAX_LEN + 1);
+	if (resp->device->ctrlr->pci_addr == NULL)
+		return -DER_NOMEM;
+	if ((req->ids == NULL) || (strlen(req->ids) == 0)) {
```
We should be avoiding strlen. Could we use strnlen with ADDR_STR_MAX_LEN?

tanabarr replied:
I will address when I have to re-push (after the parent lands).

kjacque commented on src/control/server/instance_storage_rpc.go:
```diff
@@ -238,67 +245,92 @@ func scanEngineBdevsOverDrpc(ctx context.Context, ei *EngineInstance, pbReq *ctl
+func bdevScanEngineAssigned(ctx context.Context, engine Engine, pbReq *ctlpb.ScanNvmeReq, devList *storage.BdevDeviceList, isStarted *bool) (*ctlpb.ScanNvmeResp, error) {
+	*isStarted = engine.IsStarted()
+	if !*isStarted {
+		//ei.log.Debugf("scanning engine-%d bdev tiers while engine is down", engine.Index())
```
Need to remove? Could make it a trace level log if it's useful sometimes.

tanabarr replied:
I will address when I have to re-push (after the parent lands).

kjacque reviewed (approved):
Minor things, could be addressed in a follow on.

daosbuild1 commented:
Test stage Functional Hardware Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13385/11/execution/node/1332/log

daosbuild1 commented:
Test stage Functional Hardware Large completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-13385/11/testReport/

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Functional Hardware Large completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-13385/13/testReport/

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Functional Hardware Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13385/15/execution/node/1371/log

daosbuild1 commented:
Test stage Functional Hardware Large completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-13385/15/testReport/

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Functional on EL 8.8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13385/16/execution/node/1163/log

daosbuild1 commented:
Test stage Build on Leap 15.5 with Intel-C and TARGET_PREFIX completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13385/17/execution/node/315/log

daosbuild1 commented:
Test stage Build RPM on EL 9 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13385/17/execution/node/366/log

daosbuild1 commented:
Test stage Build RPM on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13385/17/execution/node/302/log

daosbuild1 commented:
Test stage Build DEB on Ubuntu 20.04 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13385/17/execution/node/288/log

tanabarr commented:
@kjacque @knard-intel @mjmac I have rebased on master now the parent PR landed. The only changes since last review approvals are the 2 most recent commits. Other that the code is identical and conflicts have just been resolved.
Please review the most recent commits, the main change is resolving the VMD format failures which were as a result of removing the cache and not setting VMD-enabled flag. The solution involves passing the controller info explicitly and passing the enabled setting from harness storage provider to engine storage provider on engine creation.

kjacque reviewed (approved):
Overall LGTM. One minor comment, could be addressed in a follow-on.

daosbuild1 commented:
Test stage Functional Hardware Large completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-13385/20/testReport/

tanabarr commented on src/control/server/instance_storage_rpc.go:
```diff
@@ -78,11 +79,28 @@ func (ei *EngineInstance) scmFormat(force bool) (*ctlpb.ScmMountResult, error) {
 
-func (ei *EngineInstance) bdevFormat() (results proto.NvmeControllerResults) {
+func formatEngineBdevs(engine Engine, ctrlrs storage.NvmeControllers) (results proto.NvmeControllerResults) {
+	ei := engine.(*EngineInstance)
```
I will update in https://github.com/daos-stack/daos/pull/13463 when I rebase.

tanabarr commented:
CI failures all attributable to known issue https://daosio.atlassian.net/browse/DAOS-14718

mjmac reviewed (commented):
The code looks fine overall, good to ultimately simplify the logic by avoiding the cache. I see that this was tested with `Features: control` ... Is that sufficient to have a high level of confidence that this won't introduce regressions on real hardware? Does that testing include VMD-enabled/disabled configurations and md-on-ssd?

tanabarr commented:
> The code looks fine overall, good to ultimately simplify the logic by avoiding the cache. I see that this was tested with `Features: control` ... Is that sufficient to have a high level of confidence that this won't introduce regressions on real hardware? Does that testing include VMD-enabled/disabled configurations and md-on-ssd?


CI run https://build.hpdd.intel.com/blue/organizations/jenkins/daos-stack%2Fdaos/detail/PR-13385/20/pipeline/1200/ ran hardware lage, medium and medium+verbs. Medium runs with VMD AFAIK but we currently have MD-on-SSD disabled to reduce CI load. I will address your code comment and rerun with `Features: control` and `Test-nvme: auto_md_on_ssd`.

tanabarr commented on src/control/cmd/dmg/pretty/storage_nvme.go:
```diff
@@ -292,6 +292,10 @@ func PrintNvmeMetaMap(hsm control.HostStorageMap, out io.Writer, opts ...PrintCo
 					iw1 := txtfmt.NewIndentWriter(iw)
+
+					// Attach parent controller details to SMD before printing.
+					device.Ctrlr = *controller
```
done

daosbuild1 commented:
Test stage Functional Hardware Medium Verbs Provider completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-13385/21/execution/node/1364/log

daosbuild1 commented:
Test stage Functional Hardware Large completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-13385/21/testReport/

daosbuild1 commented:
Test stage Functional Hardware Large completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-13385/22/testReport/

kjacque commented on src/control/server/ctl_storage_rpc.go:
```diff
@@ -855,7 +855,8 @@ func formatNvme(ctx context.Context, req formatNvmeReq, resp *ctlpb.StorageForma
 
 		// SCM formatted correctly on this instance, format NVMe
-		cResults := formatEngineBdevs(engine, ctrlrs)
+		cResults := formatEngineBdevs(engine.(*EngineInstance), ctrlrs)
```
Do we have reasons to be positive about this type underlying the interface? I know this is the type we actually have at runtime, but IMO we ought to check if the conversion worked, and return an error (or skip?) if not.

tanabarr replied:
EngineInstance is the only type that satisfies the Engine interface I think this is okay

kjacque replied:
I do think the type ought to be checked anyway -- but I'm okay with it being addressed in a follow-on PR.

tanabarr replied:
PR-13463 is based on this PR branch and will need re-basing once this lands, I will add the check to that. Thanks Kris

kjacque reviewed (commented):
Only one comment, all other of my issues are resolved.

tanabarr commented:
@daos-stack/daos-gatekeeper I think this PR is ready for landing now thank you
